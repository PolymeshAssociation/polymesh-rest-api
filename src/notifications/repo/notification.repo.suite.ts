/* istanbul ignore file */

import { NotificationRepo } from '~/notifications/repo/notifications.repo';
import { NotificationStatus } from '~/notifications/types';

export const testNotificationRepo = async (notificationRepo: NotificationRepo): Promise<void> => {
  let id: number;

  describe('method: create', () => {
    it('should create a notification', async () => {
      const notification = await notificationRepo.create({
        status: NotificationStatus.Active,
        eventId: 0,
        subscriptionId: 1,
        nonce: 1,
        triesLeft: 3,
      });
      id = notification.id;

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { createdAt: _, ...params } = notification;
      expect(params).toMatchSnapshot();
    });
  });

  describe('method: findById', () => {
    it('should find the created notification', async () => {
      const notification = await notificationRepo.findById(id);

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { createdAt: _, ...params } = notification!;

      expect(params).toMatchSnapshot();
    });
  });

  describe('method: update', () => {
    it('should update the notification', async () => {
      await notificationRepo.update(id, { status: NotificationStatus.Acknowledged });

      const notification = await notificationRepo.findById(id);

      expect(notification?.status).toEqual(NotificationStatus.Acknowledged);
    });
  });
};
