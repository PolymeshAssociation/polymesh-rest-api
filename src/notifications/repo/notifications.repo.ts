import { NotificationModel } from '~/notifications/model/notification.model';
import { testNotificationRepo } from '~/notifications/repo/notification.repo.suite';
import { NotificationParams } from '~/notifications/types';

export abstract class NotificationRepo {
  public static readonly type = 'Notification';

  public abstract create(params: NotificationParams): Promise<NotificationModel>;
  public abstract findById(id: number): Promise<NotificationModel | undefined>;
  public abstract update(
    id: number,
    params: Partial<NotificationParams>
  ): Promise<NotificationModel>;

  /**
   * a set of tests implementers should pass
   */
  public static async test(repo: NotificationRepo): Promise<void> {
    return testNotificationRepo(repo);
  }
}
