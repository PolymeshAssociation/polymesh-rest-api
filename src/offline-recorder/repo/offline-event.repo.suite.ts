import { TopicName } from '~/common/utils/amqp';
import { OfflineEventModel } from '~/offline-recorder/model/offline-event.model';
import { OfflineEventRepo } from '~/offline-recorder/repo/offline-event.repo';

const name = TopicName.Signatures;
const body = { id: 'abc', memo: 'offline suite test' };

export const testOfflineEventRepo = async (offlineRepo: OfflineEventRepo): Promise<void> => {
  let event: OfflineEventModel;

  describe('method: recordEvent', () => {
    it('should record an event', async () => {
      event = await offlineRepo.recordEvent(name, body);
      expect(event).toMatchSnapshot();
    });
  });
};
