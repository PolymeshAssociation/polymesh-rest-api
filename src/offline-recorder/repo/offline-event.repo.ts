import { AnyModel } from '~/offline-recorder/model/any.model';
import { OfflineEventModel } from '~/offline-recorder/model/offline-event.model';
import { testOfflineEventRepo } from '~/offline-recorder/repo/offline-event.repo.suite';

export abstract class OfflineEventRepo {
  public static type = 'OfflineEvent';

  public abstract recordEvent(name: string, body: AnyModel): Promise<OfflineEventModel>;

  /**
   * a set of tests implementers should pass
   */
  public static async test(repo: OfflineEventRepo): Promise<void> {
    return testOfflineEventRepo(repo);
  }
}
