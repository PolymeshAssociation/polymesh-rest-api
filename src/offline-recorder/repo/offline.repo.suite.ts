import { TopicName } from '~/common/utils/amqp';
import { OfflineEventModel } from '~/offline-recorder/model/event.model';
import { OfflineRepo } from '~/offline-recorder/repo/offline.repo';

const name = TopicName.Submissions;
const body = { id: 'abc' };

export const testOfflineRepo = async (offlineRepo: OfflineRepo): Promise<void> => {
  let event: OfflineEventModel;

  describe('method: recordEvent', () => {
    it('should create a user', async () => {
      event = await offlineRepo.recordEvent(name, body);
      expect(event).toMatchSnapshot();
    });
  });
};
