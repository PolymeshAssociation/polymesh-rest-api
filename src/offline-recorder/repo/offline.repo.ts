import { OfflineEventModel } from '~/offline-recorder/model/event.model';
import { testOfflineRepo } from '~/offline-recorder/repo/offline.repo.suite';

export abstract class OfflineRepo {
  public static type = 'Offline';

  public abstract recordEvent(
    name: string,
    body: Record<string, unknown>
  ): Promise<OfflineEventModel>;

  /**
   * a set of tests implementers should pass
   */
  public static async test(repo: OfflineRepo): Promise<void> {
    return testOfflineRepo(repo);
  }
}
