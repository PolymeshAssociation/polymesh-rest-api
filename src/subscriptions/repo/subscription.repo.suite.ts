/* istanbul ignore file */

import { AppNotFoundError } from '~/common/errors';
import { EventType } from '~/events/types';
import { SubscriptionRepo } from '~/subscriptions/repo/subscription.repo';
import { SubscriptionStatus } from '~/subscriptions/types';

export const testSubscriptionRepo = async (subscriptionRepo: SubscriptionRepo): Promise<void> => {
  let id: number;

  describe('method: create', () => {
    it('should create a subscription', async () => {
      const subscription = await subscriptionRepo.create({
        status: SubscriptionStatus.Inactive,
        eventType: EventType.TransactionUpdate,
        eventScope: '',
        webhookUrl: 'http://example.com',
        legitimacySecret: 'someSecret',
        ttl: 10,
        triesLeft: 10,
        nextNonce: 3,
        createdAt: new Date('1987-10-14'),
      });

      id = subscription.id;
      expect(subscription).toMatchSnapshot();
    });
  });

  describe('method: findById', () => {
    it('should find the created notification', async () => {
      const subscription = await subscriptionRepo.findById(id);

      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { createdAt: _, ...params } = subscription!;

      expect(params).toMatchSnapshot();
    });

    it('should throw an error for an ID that does not exist', async () => {
      return expect(subscriptionRepo.findById(-1)).rejects.toThrow(AppNotFoundError);
    });
  });

  describe('method: update', () => {
    it('should update the notification', async () => {
      await subscriptionRepo.update(id, { status: SubscriptionStatus.Active });

      const subscription = await subscriptionRepo.findById(id);

      expect(subscription?.status).toEqual(SubscriptionStatus.Active);
    });
  });

  describe('method: findAll', () => {
    it('should return all of the subscriptions', async () => {
      const subscriptions = await subscriptionRepo.findAll();

      expect(subscriptions.length).toEqual(1);
      expect(subscriptions[0].id).toEqual(id);
    });
  });

  describe('method: incrementNonces', () => {
    it('should increment nonces', async () => {
      await subscriptionRepo.incrementNonces([id]);

      const subscription = await subscriptionRepo.findById(id);

      expect(subscription?.nextNonce).toEqual(4);
    });
  });
};
