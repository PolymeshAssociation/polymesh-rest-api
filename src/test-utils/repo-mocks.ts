/* istanbul ignore file */

import { createMock } from '@golevelup/ts-jest';
import { ValueProvider } from '@nestjs/common';

import { ApiKeyRepo } from '~/auth/repos/api-key.repo';
import { UsersRepo } from '~/users/repo/user.repo';

export const mockApiKeyRepoProvider: ValueProvider<ApiKeyRepo> = {
  provide: ApiKeyRepo,
  useValue: createMock<ApiKeyRepo>(),
};

export const mockUserRepoProvider: ValueProvider<UsersRepo> = {
  provide: UsersRepo,
  useValue: createMock<UsersRepo>(),
};

export class MockPostgresApiRepository {
  create = jest.fn();
  findOneBy = jest.fn();
  delete = jest.fn();
  save = jest.fn();
}

export class MockPostgresUserRepository {
  findOneBy = jest.fn();
  save = jest.fn();
  delete = jest.fn();
  create = jest.fn();
}
