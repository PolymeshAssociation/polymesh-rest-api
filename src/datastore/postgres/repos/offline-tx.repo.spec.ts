import { when } from 'jest-when';
import { Repository } from 'typeorm';

import { OfflineTx } from '~/datastore/postgres/entities/offline-tx.entity';
import { PostgresOfflineTxRepo } from '~/datastore/postgres/repos/offline-tx.repo';
import { OfflineTxRepo } from '~/offline-submitter/repos/offline-tx.repo';
import { MockPostgresRepository } from '~/test-utils/repo-mocks';

describe(`PostgresOfflineTxRepo ${OfflineTxRepo.type} test suite`, () => {
  const mockRepository = new MockPostgresRepository();
  const repo = new PostgresOfflineTxRepo(mockRepository as unknown as Repository<OfflineTx>);

  mockRepository.create.mockImplementation(tx => tx);

  mockRepository.save.mockImplementation(async tx => {
    when(mockRepository.findOneBy).calledWith({ id: tx.id }).mockResolvedValue(tx);
  });

  OfflineTxRepo.test(repo);
});
