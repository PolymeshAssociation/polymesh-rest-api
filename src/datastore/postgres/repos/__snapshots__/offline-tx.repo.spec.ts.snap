// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostgresOfflineTxRepo OfflineTxRepo test suite method: createTx should record the transaction request 1`] = `
OfflineTxModel {
  "address": "someAddress",
  "id": "someTestSuiteId",
  "nonce": 1,
  "payload": {
    "metadata": {
      "memo": "test utils payload",
    },
    "method": "0x01",
    "payload": {
      "address": "address",
      "blockHash": "0x01",
      "blockNumber": "-1",
      "era": "0x01",
      "genesisHash": "0x01",
      "method": "testMethod",
      "nonce": "0x01",
      "signedExtensions": [],
      "specVersion": "0x01",
      "tip": "0x00",
      "transactionVersion": "0x01",
      "version": 1,
    },
    "rawPayload": {
      "address": "address",
      "data": "0x01",
      "type": "bytes",
    },
  },
  "signature": "0x01",
  "status": "Signed",
}
`;
